lockc/src/main.rs

```diff
@@ -0,0 +1,179 @@
+                    Err(e) => error!("could not send a response for delete_container eBPF command to fanotify thread {:?}", e),
+                }
+            }
+            EbpfCommand::AddProcess {
+                container_id,
+                pid,
+                responder_tx,
+            } => {
+                let res = add_process(&mut bpf, container_id, pid);
+                match responder_tx.send(res) {
+                    Ok(_) => {},
+                    Err(e) => error!("could not send a response for add_process eBPF command to fanotify thread {:?}", e),
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn main() -> Result<()> {
+    let log_level = match env::var("LOCKC_DEBUG") {
+        Ok(_) => LevelFilter::Debug,
+        Err(_) => LevelFilter::Info,
+    };
+    TermLogger::init(
+        LevelFilter::Debug,
+        ConfigBuilder::new()
+            .set_target_level(log_level)
+            .set_location_level(log_level)
+            .build(),
+        TerminalMode::Mixed,
+        ColorChoice::Auto,
+    )?;
+
+    // Step 1: Create a synchronous thread which takes care of fanotify
+    // polling on runc binaries. We monitor all possible runc binaries to get
+    // all runc execution events (and therefore - all operations on
+    // containers).
+    // This thread has to be synchronous and cannot be a part of Tokio runtime,
+    // because it:
+    // * uses the poll() function
+    // * blocks the filesystem operations on monitored files
+    // * in case of monitoring runc, we have to be sure that we register a new
+    //   container exactly before we allow runc to be actually executed;
+    //   otherwise we cannot guarantee that lockc will actually enforce
+    //   anything on that container.
+
+    // Fanotify thread bootstrap channel - used later to start the real bootstrap
+    // of the thread. We want to bootstrap it later, after loading eBPF
+    // programs (which happens in async code in Tokio runtime).
+    let (fanotify_bootstrap_tx, fanotify_bootstrap_rx) = oneshot::channel::<()>();
+
+    // eBPF thread channel - used by fanotify thread to request eBFP operations
+    // from the async eBPF thread.
+    let (ebpf_tx, ebpf_rx) = mpsc::channel::<EbpfCommand>(32);
+
+    // Start the thread (but it's going to wait for bootstrap).
+    let fanotify_thread = thread::spawn(move || fanotify(fanotify_bootstrap_rx, ebpf_tx));
+
+    // Step 2: Setup a Tokio runtime for asynchronous part of lockc, which#
+    // takes care of:
+    // * loading and attaching of eBPF programs
+    // * fetching events/logs from eBPF programs
+    // After initializing the eBPF world, the thread from the step 1 is going
+    // to be bootstraped.
+
+    let rt = Runtime::new()?;
+
+    rt.block_on(ebpf(fanotify_bootstrap_tx, ebpf_rx))?;
+
+    // TODO(vadorovsky): Can we somehow just do `?` here, without that
+    // stupid wrapping and logging?
+    if let Err(e) = fanotify_thread.join() {
```

Comment:
Dunno, just typing that out of my head... the rust compiler will probably yell at you:

```rust
fanotify_thread.join().or_else(|e| {
  error!("boom: {}", e);
  e
})?
```

Reply:
Unfortunately that doesn't work. The error type is `Box<dyn Any + Send>` and I'm receiving an error `doesn't have a size known at compile-time`

It seems like the most of people just do `.join().unwrap()`. I will stick to what I have already.

Reply:
Sure, no worries